common/flatpak-prune.c

```diff
@@ -0,0 +1,826 @@
+
+      /* Just return if the parent isn't found; we do expect most
+       * people to have partial repositories.
+       */
+      if (commit == NULL)
+        break;
+
+      g_hash_table_add (inout_checksums, g_strdup (commit_checksum));
+
+      gboolean recurse = FALSE;
+      if (maxdepth == -1 || maxdepth > 0)
+        {
+          g_free (tmp_checksum);
+          tmp_checksum = ostree_commit_get_parent (commit);
+          if (tmp_checksum)
+            {
+              commit_checksum = tmp_checksum;
+              if (maxdepth > 0)
+                maxdepth -= 1;
+              recurse = TRUE;
+            }
+        }
+      if (!recurse)
+        break;
+    }
+
+  return TRUE;
+}
+
+/* We need to keep track of possibly a lot of object names (flathub has > 16 million objects atm),
+ * so the list of reachable objectnames need to be very compact. To handle this we use a fixed
+ * size array to reference the object names. The first 32 bytes is the checksum in raw form and
+ * the final byte is the object type.
+ */
+
+#define FLATPAK_OSTREE_OBJECT_NAME_LEN (32 + 1)
+typedef guint8 FlatpakOstreeObjectName[FLATPAK_OSTREE_OBJECT_NAME_LEN];
+
+#define FLATPAK_OSTREE_OBJECT_NAME_ELEMENT_TYPE "(yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy)" /* 32 + 1 bytes, is a fixed type */
+
+static void
+flatpak_ostree_object_name_serialize (FlatpakOstreeObjectName *name,
+                                      const char *checksum,
+                                      OstreeObjectType objtype)
+{
+  ostree_checksum_inplace_to_bytes (checksum, &(*name)[0]);
+  g_assert (objtype < 255);
+  (*name)[32] = (guint8) objtype;
+}
+
+static gint
+flatpak_ostree_name_compare (const FlatpakOstreeObjectName *name_a,
+                             const FlatpakOstreeObjectName *name_b)
+{
+  return memcmp (name_a, name_b, sizeof (FlatpakOstreeObjectName));
+}
+
+static guint
+flatpak_ostree_object_name_hash (gconstpointer a)
+{
+  const FlatpakOstreeObjectName *name = a;
+  const guint8 *data = &(*name)[0];
+
+  /* The checksum is essentially all random, so any 4 bytes of it should
+     be a good hash value. However, we avoid using the first ones, because
+     those are the ones that will be first compared on a hash collision,
+     so if they were always the same that would waste 4 comparisons. */
+  return
+    data[32] |
+    data[31] << 8 |
+    data[30] << 16 |
+    data[30] << 24;
+}
+
+static gboolean
+flatpak_ostree_object_name_equal (gconstpointer a,
+                                  gconstpointer b)
+{
+  const FlatpakOstreeObjectName *name_a = a;
+  const FlatpakOstreeObjectName *name_b = b;
+
+  return flatpak_ostree_name_compare (name_a, name_b) == 0;
+}
+
+/* This is a container for allocating FlatpakOstreeObjectNames in chunks without relocations so
+   that the resulting pointers are stable and can be stored in e.g. a hashtable. */
+
+#define BAG_CHUNK_SIZE 1985 /* nr of objects per chunk in bag, makes chunk fit in 64k with some spare for overhead */
```

Comment:
should mention that this is useful for preventing fragmentation, especially if you're going to squash the commit message with that info in it